plugins/discourse-presence/assets/javascripts/discourse/lib/presence-manager.js.es6

```diff
@@ -0,0 +1,197 @@
+import EmberObject from "@ember/object";
+import { debounce, throttle } from "@ember/runloop";
+import { ajax } from "discourse/lib/ajax";
+import computed, { observes } from "discourse-common/utils/decorators";
+
+// The durations chosen here determines the accuracy of the presence feature.
+// However, it comes at a tradeoff where decreasing the duration increases the
+// accuracy at the expense of more network calls to publish the client's state.
+//
+// Logic walk through of our heuristic implementation:
+// - When client A is typing, a message is published every KEEP_ALIVE_DURATION_SECONDS.
+// - Client B receives the message and sets a debounce of BUFFER_DURATION_SECONDS to clear
+//   all users state.
+// - Each user stored in an array on client B is marked by client B with a timestamp of when the user was seen.
+// - If client A continues to type, client B will continue to receive messages to update the timestamp
+//   of when client A was last seen.
+// - If client A disconnects or becomes inactive, the state of client A will be cleaned up in client B in two ways:
+//     * When no new replying messages have been received for BUFFER_DURATION_SECONDS which means the debounced function to clear
+//       up state on client B will be called.
+//     * When replying messages are received, client B will update the state of the new user that is replying and
+//       remove users that are stale.
+const KEEP_ALIVE_DURATION_SECONDS = 10;
+const BUFFER_DURATION_SECONDS = KEEP_ALIVE_DURATION_SECONDS + 3;
+
+export const REPLYING = "replying";
+export const EDITING = "editing";
+export const CLOSED = "closed";
+
+const PresenceManager = EmberObject.extend({
+  users: null,
+  editingUsers: null,
+  subscribed: null,
+  topic: null,
+  currentUser: null,
+  messageBus: null,
+  siteSettings: null,
+  presenceUsers: null,
+  editingPresenceUsers: null,
+
+  init() {
+    this._super(...arguments);
+
+    this.setProperties({
+      users: [],
+      editingUsers: [],
+      presenceUsers: [],
+      editingPresenceUsers: [],
+      subscribed: false
+    });
+  },
+
+  subscribe() {
+    if (this.get("subscribed")) return;
+
+    this.get("messageBus").subscribe(this.get("channel"), message => {
+      const { user, state } = message;
+      if (this.get("currentUser.id") === user.id) return;
+
+      switch (state) {
+        case REPLYING:
+          this._appendUser(this.get("users"), user);
+          this._debounceClearUsers("users");
+          break;
+        case EDITING:
+          this._appendUser(this.get("editingUsers"), user, {
+            post_id: parseInt(message.post_id, 10)
+          });
+
+          this._debounceClearUsers("editingUsers");
+          break;
+        case CLOSED:
+          this._removeUser(user);
+          break;
+      }
+    });
+
+    this.set("subscribed", true);
+  },
+
+  unsubscribe() {
+    this.get("messageBus").unsubscribe(this.get("channel"));
```

Comment:
I won't report all of them, but there's mostly no reason to do this.get("xxx") instead of this.xxx, get is only usefull when you have multiple nesting levels: this.get("foo.bar.baz")